Update compose to version 1.7.0

schott12521 opened the pull request:
- Update to compose 1.7.0-rc01
- Remove deprecated calls from upgrading compose
- Simplified iOS build by removing pods and instead relying on a "Compile Kotlin Framework" Build Phase
- Tested on iOS and Android
- Separated out build steps, so the main project can be built via `./gradlew build` and it won't build the example project, which can be built via command line or via IDE (recommended)
- Update to version `0.2.0-alpha05`
- Slimmed version from my last PR!

schott12521 commented:
If you're impatient you can grab my released version here: https://github.com/schott12521?tab=packages&repo_name=compose-cupertino

its roughly the same, but no guarantees anything works and I'll take it down as soon as this main repo is updated.

schott12521 commented:
@alexzhirkevich any feedback?

alexzhirkevich commented:
Sorry, been a busy days. I'll look throught this in the next couple days, may be on weekend  scince stable 1.7 Compose is out

alexzhirkevich commented on cupertino-native/src/iosMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoContextMenu.kt:
```diff
@@ -1,100 +0,0 @@
-///*
```
If the file is commented out it doesn't mean that it should be deleted

schott12521 replied:
Why? You're using version control, if you need the commented out code, then find it in the repo history.

This library is riddled with commented-out and unused code, making it harder to debug and find things.

Checking in commented out code is a nasty code smell that fails to benefit the library.

alexzhirkevich commented on cupertino-native/src/iosMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoDatePickerNative.ios.kt:
```diff
@@ -124,7 +125,10 @@ internal fun CupertinoDatePickerNativeImpl(
             (it.subviews.firstOrNull() as UIView?)?.backgroundColor = containerColor.toUIColor()
             it.applyTheme(dark)
         },
-        background = containerColor
```
Doesn't new UIKitView overload support background color?

schott12521 replied:
No it does not, but I didn't notice any loss of features when removing this:

```
/*
 * Copyright 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.compose.ui.viewinterop

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import platform.UIKit.UIView

/**
 * Compose a [UIView] of class [T] into the UI hierarchy.
 *
 * @param factory The block creating the [T] to be composed.
 *
 * NOTE: [T] shouldn't be leaked outside and will be managed by Compose runtime efficiently.
 * Remembering [T] externally and passing it to be returned from [factory] can (and probably will)
 * lead to hilarious bugs in case [onReset] is not `null`.
 * @param modifier The modifier to be applied to the layout.
 * @param update A callback to be invoked every time the state it reads changes.
 * Invoked once initially and then every time the state it reads changes.
 * @param onRelease A callback invoked as a signal that the [T] has exited the
 * composition forever. Use it to release resources and stop jobs associated with [T].
 * @param onReset If not null, this callback is invoked when this composable node is
 * reused in the composition instead of being recreated. Use it to reset the state of [T] to
 * some blank state. This is a function that will be executed instead of [factory] if the node
 * containing [T] was reused. If null, [T] will not be reused, a new instance of [T] will be created
 * using [factory] every time this function enters the composition.
 * @property properties The properties configuring the behavior of [T]. Default value is
 * [UIKitInteropProperties.Default]
 *
 * @see UIKitInteropProperties
 */
@Composable
fun <T : UIView> UIKitView(
    factory: () -> T,
    modifier: Modifier = Modifier,
    update: (T) -> Unit = NoOp,
    onRelease: (T) -> Unit = NoOp,
    onReset: ((T) -> Unit)? = null,
    properties: UIKitInteropProperties = UIKitInteropProperties.Default,
) {
    val interopContainer = LocalInteropContainer.current

    InteropView(
        factory = { compositeKeyHash ->
            UIKitInteropViewHolder(
                factory,
                interopContainer,
                properties,
                compositeKeyHash,
            )
        },
        modifier,
        onReset,
        onRelease,
        update = {
            update(it)

            val holder = interopContainer.holderOfView(it) as? UIKitInteropViewHolder<*>
            holder?.properties = properties
        }
    )
}
```

schott12521 commented on cupertino-native/src/iosMain/kotlin/io/github/alexzhirkevich/cupertino/PresentationController.ios.kt:
```diff
@@ -78,10 +77,6 @@ internal fun <T : UIViewController> PresentationController(
                     .animationControllerForPresentedController(presented, presentingController, sourceController)
             }
 
-            override fun interactionControllerForDismissal(animator: UIViewControllerAnimatedTransitioningProtocol): UIViewControllerInteractiveTransitioningProtocol? {
-                return currentDelegate.interactionControllerForDismissal(animator)
-            }
-
```
This was causing the Alerts to crash when they were clicked from iOS. After removing this, the Alerts work correctly without any loss in features that I can see.

schott12521 commented on cupertino-native/src/iosMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoDialogsNative.ios.kt:
```diff
@@ -140,18 +117,38 @@ internal fun UIAlertController(
     }
 }
 
+// Function to handle alert dismissal with safe flag usage
+private fun dismissAlert(onDismissRequest: () -> Unit) {
+    if (!isAlertControllerBeingDismissed) {
+        isAlertControllerBeingDismissed = true
+        onDismissRequest()
+        // Reset the flag after dismissal is complete
+        isAlertControllerBeingDismissed = false
+    }
+}
+
```
This was something I had for debugging why Dialogs were crashing, I will remove this.

schott12521 commented on cupertino-native/src/iosMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoDialogsNative.ios.kt:
```diff
@@ -168,18 +165,22 @@ private class NativeAlertDialogButtonsScopeImpl(
             handler = {
-                onClick()
-                onDismissRequest()
+                try {
+                    onClick()
+                } catch (exception: Exception) {
+                    println("We tried the onclick and there was an exception")
+                }
```
This is unused after removing the additional exception handling I added when debugging the Dialogs on iOS. Will remove in the next commit.

schott12521 commented on cupertino/src/commonMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoIndication.kt:
```diff
@@ -1,107 +0,0 @@
-/*
```
I was able to accomplish this and will include in the next commit.

schott12521 commented:
I've addressed PR feedback, including:

- Removing unnecessary changes I made to CupertinoDialogs, the crash while updating was due to this line of `PresentationController` for cupertino-native:

```
override fun interactionControllerForDismissal(animator: UIViewControllerAnimatedTransitioningProtocol): UIViewControllerInteractiveTransitioningProtocol? {
                return currentDelegate.interactionControllerForDismissal(animator)
}
```
- Refactored CupertinoIndicator to not use deprecated methods
- Updated from compose 1.7.0-rc01 to 1.7.0

Let me know if there are any other concerns or issues @alexzhirkevich

Tweener commented:
> I've addressed PR feedback, including:
> 
> * Removing unnecessary changes I made to CupertinoDialogs, the crash while updating was due to this line of `PresentationController` for cupertino-native:
> 
> ```
> override fun interactionControllerForDismissal(animator: UIViewControllerAnimatedTransitioningProtocol): UIViewControllerInteractiveTransitioningProtocol? {
>                 return currentDelegate.interactionControllerForDismissal(animator)
> }
> ```
> 
> * Refactored CupertinoIndicator to not use deprecated methods
> * Updated from compose 1.7.0-rc01 to 1.7.0
> 
> Let me know if there are any other concerns or issues @alexzhirkevich

I was just testing on Compose 1.7.0 when I saw it crashes. I came to this repo to check if a new version was available to support 1.7.0, and just saw your PR. Good timing 😁

Thank you for your work @schott12521 !
And hopefully @alexzhirkevich will be able to release a new version soon 🙂 Great lib!

alexzhirkevich reviewed (approved):
Thanks

Tweener commented:
Hey @alexzhirkevich, any chance you could make a new release with this fix?
Thanks 🙏

schott12521 commented:
@Tweener I think @alexzhirkevich is holding off until some bugs are fixed: https://github.com/alexzhirkevich/compose-cupertino/issues/55#issuecomment-2453127239

I've got a new SwipeBox implementation that is performing much better; hopefully we can merge it quickly after I make a PR. Are there any other release blocking bugs @alexzhirkevich?

Tweener commented:
@schott12521 I see, thanks for letting me know!
